File: 198_打家劫舍/rob.py
# import numpy as np

class Solution:
    def rob(self, nums: List[int]) -> int:
        # 只与前面一个状态有关
        # n = len(nums)
        # dp = np.zeros((n, 2))
        # dp[0][1] = nums[0]
        # for i in range(1, n):
        #     dp[i][0] = max(dp[i-1][0], dp[i-1][1])
        #     dp[i][1] = dp[i-1][0] + nums[i]

        # return int(max(dp[-1][0], dp[-1][1]))

        # 与前面两个状态有关
        n = len(nums)
        if n == 1:
            return nums[0]
        a, b = nums[0], max(nums[0], nums[1])
        for i in range(2, n):
            a, b = b, max(b, a + nums[i])
        return b


# if __name__ == "__main__":
#     solu = Solution()
#     nums = [0]
#     print(solu.rob(nums))

File: 198_打家劫舍/rob.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    // dp[i] 从第 i 家开始偷，能偷到最多的钱
    // let dp = new Array(nums.length+2).fill(0)
    let cur = 0; let pos1 = 0; let pos2 = 0
    let res = 0
    for(let i = nums.length-1; i >= 0; i--){
        // dp[i] = Math.max(dp[i+1], dp[i+2]+nums[i])
        cur = Math.max(pos1, pos2+nums[i])
        res = Math.max(res, cur)
        pos2 = pos1
        pos1 = cur
    }
    // return Math.max(...dp)
    return res
};